kimchi/src/circuits/lookup/lookups.rs

```diff
@@ -179,44 +179,65 @@ impl<F: Copy> SingleLookup<F> {
     }
 }
 
+/// The table ID associated with a particular lookup
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub enum LookupTableID {
+    /// Look up the value from the given fixed table ID
+    Constant(i32),
+    /// Look up the value in the table with ID given by the value in the witness column
+    WitnessColumn(usize),
+}
+
 /// A spec for checking that the given vector belongs to a vector-valued lookup table.
 #[derive(Clone, Serialize, Deserialize, Debug)]
-pub struct JointLookup<SingleLookup> {
+pub struct JointLookup<SingleLookup, LookupTableID> {
     /// The ID for the table associated with this lookup.
     /// Positive IDs are intended to be used for the fixed tables associated with individual gates,
     /// with negative IDs reserved for gates defined by the particular constraint system to avoid
     /// accidental collisions.
-    pub table_id: i32,
+    pub table_id: LookupTableID,
     pub entry: Vec<SingleLookup>,
 }
 
 /// A spec for checking that the given vector belongs to a vector-valued lookup table, where the
 /// components of the vector are computed from a linear combination of locally-accessible cells.
-pub type JointLookupSpec<F> = JointLookup<SingleLookup<F>>;
+pub type JointLookupSpec<F> = JointLookup<SingleLookup<F>, LookupTableID>;
+
+/// A concrete value or representation of a lookup.
+pub type JointLookupValue<F> = JointLookup<F, F>;
 
-impl<F: Zero + One + Clone + Neg<Output = F> + From<u64>> JointLookup<F> {
+impl<F: Zero + One + Clone + Neg<Output = F> + From<u64>> JointLookupValue<F> {
     // TODO: Support multiple tables
     /// Evaluate the combined value of a joint-lookup.
     pub fn evaluate(&self, joint_combiner: &F, table_id_combiner: &F) -> F {
         combine_table_entry(
             joint_combiner,
             table_id_combiner,
             self.entry.iter(),
-            i32_to_field(self.table_id),
+            &self.table_id,
         )
     }
 }
 
-impl<F: Copy> JointLookup<SingleLookup<F>> {
+impl<F: Copy> JointLookup<SingleLookup<F>, LookupTableID> {
     /// Reduce linear combinations in the lookup entries to a single value, resolving local
     /// positions using the given function.
-    pub fn reduce<K, G: Fn(LocalPosition) -> K>(&self, eval: &G) -> JointLookup<K>
+    pub fn reduce<K, G: Fn(LocalPosition) -> K>(&self, eval: &G) -> JointLookupValue<K>
```

Comment:
I was really confused as to why this was taking a closure instead of writing the code without a closure, looking at this again I'm confused again

Also why can't `K` be the `F` here?

Reply:
`K` is a vector space over `F`, which allows us to mix expr types with field constants using the same code.